pkg/controllers/reallocation/suite_test.go

```diff
@@ -122,5 +151,64 @@ var _ = Describe("Reallocation", func() {
 			Expect(updatedNode.Labels).ToNot(HaveKey(v1alpha3.ProvisionerUnderutilizedLabelKey))
 			Expect(updatedNode.Annotations).ToNot(HaveKey(v1alpha3.ProvisionerTTLAfterEmptyKey))
 		})
+		It("should terminate underutilized nodes past their TTL", func() {
+			node := test.Node(test.NodeOptions{
+				Finalizers: []string{v1alpha3.KarpenterFinalizer},
+				Labels: map[string]string{
+					v1alpha3.ProvisionerNameLabelKey:          provisioner.Name,
+					v1alpha3.ProvisionerUnderutilizedLabelKey: "true",
+				},
+				Annotations: map[string]string{
+					v1alpha3.ProvisionerTTLAfterEmptyKey: time.Now().Add(-100 * time.Second).Format(time.RFC3339),
+				},
+			})
+			ExpectCreated(env.Client, provisioner, node)
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(provisioner))
+
+			updatedNode := &v1.Node{}
+			Expect(env.Client.Get(ctx, client.ObjectKey{Name: node.Name}, updatedNode)).To(Succeed())
+			Expect(updatedNode.DeletionTimestamp.IsZero()).To(BeFalse())
+		})
+		It("should only terminate nodes that failed to join with all pods terminating after 5 minutes", func() {
+			node := test.Node(test.NodeOptions{
+				Finalizers: []string{v1alpha3.KarpenterFinalizer},
+				Labels: map[string]string{
+					v1alpha3.ProvisionerNameLabelKey:          provisioner.Name,
+					v1alpha3.ProvisionerUnderutilizedLabelKey: "true",
+				},
+				ReadyStatus: v1.ConditionUnknown,
+			})
+			pod := test.Pod(test.PodOptions{
+				Finalizers: []string{"fake.sh/finalizer"},
+				NodeName:   node.Name,
+			})
+			ExpectCreated(env.Client, provisioner, pod)
+			ExpectCreatedWithStatus(env.Client, node)
+
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(provisioner))
+
+			// Expect node not deleted
+			updatedNode := ExpectNodeExists(env.Client, node.Name)
+			Expect(updatedNode.DeletionTimestamp.IsZero()).To(BeTrue())
+
+			// Set pod DeletionTimestamp and do another reconcile
+			Expect(env.Client.Delete(ctx, pod)).To(Succeed())
+			ExpectReconcileSucceeded(ctx, controller, client.ObjectKeyFromObject(provisioner))
+
+			// Expect node not deleted
+			updatedNode = ExpectNodeExists(env.Client, node.Name)
+			Expect(updatedNode.DeletionTimestamp.IsZero()).To(BeTrue())
+
+			future := time.Now().Add(500 * time.Second)
+			// Monkey patch time.Now() so it returns 500 seconds in the future
+			monkey.Patch(time.Now, func() time.Time {
```

Comment:
consider `monkey.Patch(time.Now, func() time.Time { time.Now().Add(500 * time.Second) })`

Reply:
Can't reference `time.Now` while we patch `time.Now`'s function address. This results in a buffer overflow.

Reply:
Interesting! Still, you could inline 204-206.